Move doc comment parsing to rustc_lexer

matklad opened the pull request:
Plain comments are trivia, while doc comments are not, so it feels
like this belongs to the rustc_lexer.

The specific reason to do this is the desire to use rustc_lexer in
rustdoc for syntax highlighting, without duplicating "is this a doc
comment?" logic there.

r? @ghost

matklad committed:
Move doc comment parsing to rustc_lexer

Plain comments are trivial, while doc comments are not, so it feels
like this belongs to the rustc_lexer.

The specific reason to do this is the desire to use rustc_lexer in
rustdoc for syntax highlighting, without duplicating "is this a doc
comment?" logic there.

matklad commented:
r? @petrochenkov

matklad commented on src/librustc_parse/lexer/mod.rs:
```diff
@@ -318,6 +315,36 @@ impl<'a> StringReader<'a> {
+        comment_kind: CommentKind,
+        doc_kind: rustc_lexer::DocCommentKind,
+    ) -> TokenKind {
+        let mut idx = 0;
+        loop {
+            idx = match content[idx..].find('\r') {
+                None => break,
+                Some(it) => idx + it + 1,
+            };
+            self.err_span_(
+                content_start + BytePos(idx as u32 - 1),
+                content_start + BytePos(idx as u32),
+                match comment_kind {
+                    CommentKind::Line => "bare CR not allowed in doc-comment",
+                    CommentKind::Block => "bare CR not allowed in block doc-comment",
+                },
+            );
+        }
```
This loop makes me sad, but zero-cost abstraction are not so zero cost here: https://godbolt.org/z/vE7xsr :(

Mark-Simulacrum replied:
`char_indices`-based slow seems to at least unroll somewhat and avoids any panic-handling: https://godbolt.org/z/WP5YaP

matklad replied:
Hm, actually, we should just fast-path this...

matklad committed:
Simplify search for bare `\r` in doc comments

Outer `if` is the fast path -- it calls into hyperoptimized memchr.

The inner loop is just the simplest code possible -- it doesn't
generated the tightest code, but that shouldn't matter if we are going
to error anyhow.

petrochenkov commented on src/librustc_ast/util/comments.rs:
```diff
@@ -216,8 +176,8 @@ pub fn gather_comments(sm: &SourceMap, path: FileName, src: String) -> Vec<Comme
             }
-            rustc_lexer::TokenKind::BlockComment { terminated } => {
-                if block_doc_comment_style(token_text, terminated).is_none() {
+            rustc_lexer::TokenKind::BlockComment { kind, terminated: _ } => {
```
```suggestion
            rustc_lexer::TokenKind::BlockComment { kind, .. } => {
```

petrochenkov commented on src/librustc_ast/util/comments.rs:
```diff
@@ -216,8 +176,8 @@ pub fn gather_comments(sm: &SourceMap, path: FileName, src: String) -> Vec<Comme
-            rustc_lexer::TokenKind::BlockComment { terminated } => {
-                if block_doc_comment_style(token_text, terminated).is_none() {
+            rustc_lexer::TokenKind::BlockComment { kind, terminated: _ } => {
+                if kind == rustc_lexer::CommentKind::Plain {
```
Nit: could you `use rustc_lexer::CommentKind` instead?
Long paths like this are not normally used in rustc.

matklad replied:
Ack. I'll leave `rustc_lexer::TokenKind` paths in though, as there are two `TokenKind`s (and this is unrelated to PR)

matklad replied:
Hm, actually, let me take this back. In the `lexer/mod.rs`, if I do replacement, I get

```
    fn cook_doc_comment(
        &self,
        content_start: BytePos,
        content: &str,
        comment_kind: CommentKind,
        doc_kind: DocCommentKind,
    ) -> TokenKind {
```

The problem here is that `CommentKind` and `DocCommentKind` are from different crates, which are confusing.

petrochenkov replied:
With this PR `CommentKind` from `rustc_ast` can be eliminated.

petrochenkov replied:
In case of duplicated types `rust_lexer` version have to be qualified, yes.

matklad replied:
Hm, either I disagree, or I don't understand. The two `CommentKind`s are orthogonal axis: one is Inner/Outer, the other is Line/Block. ast's CommentKind (Line/Block) fits the current shape of the ast, it doesn't seem worth it to change it.

What can be merged is `rustc_lexer::CommentKind` and `rustc_ast::AttrStyle`, as they are both just `InnerOrOuter`. I think it's nicer than they are separate though, as they exist on different levels of abstraction, and `AttrStyle` is also serializable.

matklad replied:
Edited https://github.com/rust-lang/rust/pull/75642#issuecomment-676058846 to better show type correspondence. *Independently*, the naming is good :-(

petrochenkov replied:
Ok, I didn't have the code in front of my eyes and mixed everything up.

petrochenkov commented on src/librustc_parse/lexer/mod.rs:
```diff
@@ -318,6 +315,33 @@ impl<'a> StringReader<'a> {
             }
         }
     }
```
Nit: missing newline between functions.

petrochenkov commented on src/librustc_parse/lexer/mod.rs:
```diff
@@ -318,6 +315,33 @@ impl<'a> StringReader<'a> {
+
+        let attr_style = match doc_kind {
+            rustc_lexer::DocCommentKind::Inner => AttrStyle::Inner,
+            rustc_lexer::DocCommentKind::Outer => AttrStyle::Outer,
```
Same comment about `use rustc_lexer::DocCommentKind`.

petrochenkov commented:
r=me with nits addressed.

matklad commented:
I've addressed nits *except* for qualified paths. I think, *in this case*, qualified paths would be better. 

We have parallel hierarchies of types here:

rustc_lexer: `Token`, `TokenKind`, `CommentKind`, `DocCommentKind`
rustc_ast: `Token`, `TokenKind`, `CommentKind`

So, it seems better if we stick to not qualifying `_ast` and qualifing `_lexer`

petrochenkov commented:
```
CommentKind -----------
----------- CommentKind
```
This looks pretty bad.
I think we need to come up with a different name for one of these.

petrochenkov commented:
Block vs Line looks like a "heavier" candidate for `CommentKind` to me.

The other can be `CommentIsDoc::(Yes,No)`.

Also `DocCommentKind` -> `DocCommentStyle` (and change its variant order) to mirror `AttrStyle` closer.
I'd probably just put `AttrStyle` as is into the lexer, personally, but there are issues with (en,de)coding and hash impls.

matklad commented:
How this version look? I've get rid of `CommentKind` and renamed `DocCommentKind -> DocStyle`

petrochenkov commented:
LGTM.
@bors r+

bors commented:
:pushpin: Commit ccbe94bf77e6a32fc9f31425bc820345be3143c0 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov ccbe94bf77e6a32fc9f31425bc820345be3143c0 -->
<!-- homu: {"type":"Approved","sha":"ccbe94bf77e6a32fc9f31425bc820345be3143c0","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit ccbe94bf77e6a32fc9f31425bc820345be3143c0 with merge b51651ae9d0161967617be930415705b2e4d5faf...
<!-- homu: {"type":"BuildStarted","head_sha":"ccbe94bf77e6a32fc9f31425bc820345be3143c0","merge_sha":"b51651ae9d0161967617be930415705b2e4d5faf"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1011330408), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36130)
Approved by: petrochenkov
Pushing b51651ae9d0161967617be930415705b2e4d5faf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36130","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1011330408"},"merge_sha":"b51651ae9d0161967617be930415705b2e4d5faf"} -->
